lib/dest/writeContents/index.js

```diff
@@ -29,49 +41,124 @@ function writeContents(writePath, file, cb) {
 
   // If no contents then do nothing
   if (file.isNull()) {
-    return complete();
+    return finish();
   }
 
-  function complete(err) {
-    cb(err, file);
-  }
 
-  function written(err) {
+  // This is invoked by the various writeXxx modules when they've finished
+  // writing the contents.
+  // The third argument, if present, should be invoked to indicate we're done
+  // with the file descriptor.
+  function written(err, fd, close) {
+    close = close || function(err, cb) {
+      cb(err);
+    };
+
+    if (err && !(err.code === 'EEXIST' && file.flag === 'wx')) {
+      return close(err, finish);
+    }
 
-    if (isErrorFatal(err)) {
-      return complete(err);
+    // TODO handle symlinks properly
+    if (!file.stat || file.symlink) {
+      return close(null, finish);
     }
 
-    if (!file.stat || typeof file.stat.mode !== 'number' || file.symlink) {
-      return complete();
+    if (typeof fd === 'number') {
+      return stat(fd, close);
     }
 
-    fs.stat(writePath, function(err, st) {
+    // No file descriptor given (writeDir or writeSymbolicLink) so create one.
+    fs.open(writePath, 'r', function(err, fd) {
       if (err) {
-        return complete(err);
+        return close(err, finish);
       }
-      var currentMode = (st.mode & parseInt('0777', 8));
-      var expectedMode = (file.stat.mode & parseInt('0777', 8));
-      if (currentMode === expectedMode) {
-        return complete();
-      }
-      fs.chmod(writePath, expectedMode, complete);
+
+      stat(fd, function(err1) {
+        fs.close(fd, function(err2) {
+          finish(err1 || err2);
+        });
+      });
     });
   }
 
-  function isErrorFatal(err) {
-    if (!err) {
-      return false;
-    }
 
-    if (err.code === 'EEXIST' && file.flag === 'wx') {
-      // Handle scenario for file overwrite failures.
-      return false;   // "These aren't the droids you're looking for"
-    }
+  // Set mode and/or atime, mtime.
+  function stat(fd, close) {
+    fs.fstat(fd, function(err, stat) {
+      if (err) {
+        return close(err, finish);
+      }
+
+      // Check if mode needs to be updated
+      var modeDiff = 0;
+      if (typeof file.stat.mode === 'number') {
+        modeDiff = (file.stat.mode ^ stat.mode) & MASK_MODE;
```

Comment:
nvm, found it at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#Bitwise_XOR